### PR TITLE
Neighbors & Migration

### PR DESCRIPTION
**Pull request by mcmcgrath13**
* Migration by name or category (arbitrary category)
* Probs in csv file, can include optionally a column prob to control whether migration happens (default is 1, but that doesn't necessarily mean the agent actually moves as their current location could be selected)

To do:
- [x] what to do with migration probs file (copy from parse_params maybe?)
- [x] Neighborhood grid - I'm thinking have this be edges, but provide a utility that can create an edge list from a grid file instead (if not this approach, delete edges?)
- [x] Neighbor assorting
- [x] bump version
- [x] integration test(s)
- [x] docs

### CODE REVIEW

File: titan/params/assort_mix.yml [hunk 1]
```diff
@@ -18,7 +18,7 @@ assort_mix:
       description: "Value of agent's attribute which triggers using this rule.  Can use the value '__any__' to then allow assoring on partner_values of '__same__' and '__other__'."
       type: none
     partner_values:
-      description: "Which partners to select if assortative mixing happens. Expects a sub-key of an attribute value with a value of the probability. `__other__` can be used as an attribute value as a fallthrough option.  All of the probabilities should add to one."
+      description: "Which partners to select if assortative mixing happens. Expects a sub-key of an attribute value with a value of the probability. `__other__` can be used as an attribute value as a fallthrough option.  If the `partner_attribute` is `location`, it is also possible to use the key `__neighbor__` here to indicate a probability of assoring with agents from neighboring locations (vs. `__same__` or `__other__`).  All of the probabilities should add to one."
```
Comment (s-bessey) [inline titan/params/assort_mix.yml:21]: typo on "assorting"

File: titan/params/location.yml [hunk 2]
```diff
@@ -27,9 +27,25 @@ location:
...
+      default: name
+      type: enum
+      description: Which location attribute to determine migration probability based on.
+      values:
+        - name
+        - category
+    probs_file:
+      default: ''
+      type: any
+      description: File path where the migration probability csv is located.  The CSV must have the attribute names along the first row and column with the topmost left cell blank.  Each value in the matrix represents the probability of an agent migrating from the location on that row to the location on that column at a given time step.  Each row must add to 1.  If there is a column with the header `prob` this will be used as the probability of an agent in the location described by the row migrating at all that timestep (useful if migrating based on a non-unique attribute).
```
Comment (s-bessey) [inline titan/params/location.yml:45]: Example of this? In the params-app maybe?

File: titan/utils.py [hunk 3]
```diff
@@ -333,3 +336,129 @@ def set_up_logging(params):
...
+    """
+    Read a csv describing the layout of locations and return a dictionary describing the location edges, which can then be used in the params [location.edges].
+
+    Sample csv:
+    ```
+    location_1,location_3,location_4
+    location_1,location_3,location_4
+    location_1,location_3,location_4
+    location_1,location_2,
+    location_2,location_2,
```
Comment (s-bessey) [inline titan/utils.py:357]: Why repeat the same line three times? For the diagonal neighbors thing, but wouldn't you just need it twice for that?

File: tests/integration_test.py [hunk 4]
```diff
@@ -606,7 +618,104 @@ def test_pca_awareness(params_integration, tmpdir):
...
+    model = TITAN(params)
+
+    west_agents = [a for a in model.pop.all_agents if a.location.name == "west"]
+    num_init_west_agents = len(west_agents)
+
+    assert math.isclose(num_init_west_agents, 40, abs_tol=15)
+
+    model.run(path_a)
+
+    # most west_agents should now be east
```
Comment (s-bessey) [inline tests/integration_test.py:649]: This is unidirectional migration, right? Should it also check to make sure none of the east agents are now in the west?

File: tests/integration_test.py [hunk 5]
```diff
@@ -606,7 +618,104 @@ def test_pca_awareness(params_integration, tmpdir):
...
+
+    partners = {"east": 0, "west": 0, "north": 0, "south": 0}
+    for a in model.pop.all_agents:
+        if a.location.name == "west":
+            for p in a.iter_partners():
+                partners[p.location.name] += 1
+
+    total_partners = sum(partners.values())
+
+    # because if we have a west partner we're guaranteed 2, the data skews a bit westward TO_REVIEW, does this make sense??
```
Comment (s-bessey) [inline tests/integration_test.py:699]: I think we need to talk about this? Not sure what the comment is saying!

File: titan/utils.py (hunk 3 above)
Comment (mcmcgrath13) [reply titan/utils.py:357]: need, no.  But that's just the geography I made up in my head?   Is it confusing?

File: tests/integration_test.py (hunk 4 above)
Comment (mcmcgrath13) [reply tests/integration_test.py:649]: Not specifically, I don't think.  As over time west agents can migrate away and migrate back (except all those probs are 0 and east staying in east probability is very high).  So we could check east to west, but it seems maybe a bit redundant?

**Pull request merged**